File: frontend/src/shared/lib/unitHierarchy.ts
/** Desglose fardo → paquete → unidad (cantidad siempre en unidades base). */

export function splitStockHierarchy(
  totalUnits: number,
  unitsPerPackage: number,
  packagesPerFardo: number,
): { fardos: number; paquetes: number; unidades: number } {
  const upp = Math.max(1, Math.floor(unitsPerPackage) || 1)
  const ppf = Math.max(1, Math.floor(packagesPerFardo) || 1)
  const perFardo = upp * ppf
  const t = Math.max(0, Math.floor(totalUnits) || 0)
  const fardos = Math.floor(t / perFardo)
  const rem = t % perFardo
  const paquetes = Math.floor(rem / upp)
  const unidades = rem % upp
  return { fardos, paquetes, unidades }
}

export function formatHierarchyLabel(fardos: number, paquetes: number, unidades: number): string {
  return `${fardos} f · ${paquetes} pq · ${unidades} u`
}

/** Stock total en unidades a partir de fardos + paquetes + unidades sueltas y la jerarquía del producto. */
export function totalUnitsFromHierarchy(
  fardos: number,
  paquetes: number,
  unidadesSueltas: number,
  unitsPerPackage: number,
  packagesPerFardo: number,
): number {
  const upp = Math.max(1, Math.floor(unitsPerPackage) || 1)
  const ppf = Math.max(1, Math.floor(packagesPerFardo) || 1)
  const perFardo = upp * ppf
  const f = Math.max(0, Math.floor(fardos) || 0)
  const p = Math.max(0, Math.floor(paquetes) || 0)
  const u = Math.max(0, Math.floor(unidadesSueltas) || 0)
  return f * perFardo + p * upp + u
}

File: backend/inventory/unit_hierarchy.py
"""Desglose fardo → paquete → unidad (stock siempre en unidades base)."""


def split_stock_hierarchy(
    total_units: int,
    units_per_package: int,
    packages_per_fardo: int,
) -> tuple[int, int, int]:
    """
    Dado el total en unidades y la jerarquía del producto, devuelve (fardos, paquetes_resto, unidades_resto).
    1 fardo = packages_per_fardo paquetes; 1 paquete = units_per_package unidades.
    """
    upp = max(1, int(units_per_package or 1))
    ppf = max(1, int(packages_per_fardo or 1))
    per_fardo = upp * ppf
    t = max(0, int(total_units))
    fardos = t // per_fardo
    rem = t % per_fardo
    paquetes = rem // upp
    unidades = rem % upp
    return fardos, paquetes, unidades


def hierarchy_label(fardos: int, paquetes: int, unidades: int) -> str:
    return f'{fardos} f · {paquetes} pq · {unidades} u'
